### PR TITLE
update sortCooIndicesGeneric to take any data type

### PR DESCRIPTION
**Pull request by pza94**
Sorting indices for for COO sparse matrices was broken in by the migration to DataTypes in shapeinfo.

In current master, the data type for the values vector is hardcoded to `int64` which makes it work for only 64bit data types.

## What changes were proposed in this pull request?

use `BUILD_SINGLE_SELECTOR` for templating all possible data types and pass in data type as part of a `xShapeInfo`

## How was this patch tested?

`nd4j/linalg/specials/SortCooTests.java` modified to call sortCooIndicesGeneric with values databuffer of different data types.

## Quick checklist

The following checklist helps ensure your PR is complete:

- [x] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [x] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [x] Created tests for any significant new code additions.
- [x] Relevant tests for your changes are passing.

### CODE REVIEW

**Review by quickwritereader — DISMISSED** [review]
lgtm.

**Comment by quickwritereader** [comment]
it seems you need to add those implementation functions (the ones in NativeOps.cpp) in NaticeOps.cu to fix unresolved errors on Cuda builds. I believe you will just copy-paste the same code.
Also, please don't forget to rebase commits squashing it into one with a detailed message.
thanks. well done

**Comment by pza94** [comment]
would you like me to do something at this point, or are we just waiting for a final approval?

**Comment by quickwritereader** [comment]
> would you like me to do something at this point, or are we just waiting for a final approval?

I did not notice that you also changed the signature of the method. plus it led to an unresolved error on Cuda build

**Comment by pza94** [comment]
Changing the signature is necessary in order to pass in data type. I opted for replacing `rank` with a shape buffer rather than adding a new `dataType` parameter.

The problem with the cuda build was that the changed signature in `NativeOps.h` was not updated in `NativeOps.cu`

**Comment by pza94** [comment]
Apparently there are some tests failing in deeplearning4j-modelimport. Is there something I could do about it?

**Review by quickwritereader — APPROVED** [review]

**Review by agibsonccc — APPROVED** [review]

**Pull request merged**
